New elements - maintain your tests (#264)
OctoMind-dev/mintlify-docs: maintain-tests.mdx

---
title: Maintain your tests
description: "How to fix your broken tests easily"
icon: "traffic-cone"
---

## Why tests fail

`Red` signals the test has failed. This is due to 3 reasons:

> 1. **You have a bug in your app**

Reviewer: I think we need to get away from the generic wording that you have a bug and rather start talking about the fact that you **introduced** a bug. Something that used to work (covered by the test) is now not working anymore (what we like to call a regression in dev/product terms). Maybe even build a bridge that this is easily spotted if you produce regular test reports (scheduling *wink*) because then you actually know at which point in the time the bug got introduced (not just that it broke!)

Author: too broad of a discussion for a PR comment. I go basic with the wording for docs, for the part of the audience who has no clue what regression is, docs have to be generic a bit. And to the point, I try to keep it shorter to keep people RTFM. Suggest something if you have a snappy version in mind

<Frame caption="Snapshots in test reports showing what happened during the test run, 7/2024">
  <img
    src="/images/executing/report-snapshots.png"
    alt="snapshots carousel in test reports"
  />
</Frame>

2. **The test is broken**

End-to-end tests tend to break when the app changes. This is not just annoying, but a major time investment.

If you would like to fix the test on your own, read further. But we can also fix it for you. This feature in included in [Octomind's paid tiers](https://www.octomind.dev/pricing).

<Frame caption="Example of a broken test reflecting a change in user flow, trace viewer screenshot 7/2024">
  <img
    src="/images/executing/broken-test.png"
    alt="broken test in Playwright trace viewer"
  />
</Frame>

3. **The test is flaky**

Flaky tests produce different results when running the exact same test case. We are deploying multiple strategies to fight flakiness, but it can't be prevented 100% of the time just yet.

## Inspect the reason of the failure

> From the [test report](/account-view/test-reports.mdx), go to the **debug** tab of your test detail and launch the **trace viewer**.

Reviewer: ```suggestion
From the [test report](/account-view/test-reports.mdx), go to the **debug** tab of your test result and launch the **trace viewer**.
```

<Frame caption="Launch trace viewer for more test run details, 7/2024">
  <img
    src="/images/executing/see-trace-viewer.png"
    alt="launch trace viewer"
    width="300px"
  />
</Frame>

The **trace viewer** will highlight the exact place in the test execution where the test failed. Learn how to inspect [Playwright trace viewer](https://playwright.dev/docs/trace-viewer) step by step to identify the culprit of the broken test.
Our open source [Debugtopus](/debugtopus.mdx) is an additional help that allows you to set breakpoints and step through the code.

> ## Correct the broken test steps

Reviewer: Nitpick/ not sure how our usual wording is but my mental model is that a test **breaks** when a step **fails** to execute

Author: yes ... and wha would you like to change? is that contrary to the stuff written there?

Reviewer: Heading says "broken test case steps"

> This applies when your app is OK, but the test is broken. Use our editing tools to quickly add, remove steps or adjust locators. Here is how you [edit test cases](/edit-test-case.mdx).

Reviewer: ```suggestion
This applies when your app works as intended, but the test fails to run. Use our editing tools to quickly add, remove steps or adjust locators. Here is how you [edit test cases](/edit-test-case.mdx).
```

## The Hack: How we fix our own tests

Test often break in more places and fixing them one by one can become annoying. It doesn't have to.

Since we dogfood our own end-to-end tests, we have discovered this easy hack:

1. go to the test case

> 2. check the **prompt** - if it's not valid anymore, adjust it

Reviewer: up-to-date? matching the desired flow? what's valid in terms of prompt?

Author: ```suggestion
2. check the **prompt** - if it's not matching the desired flow anymore, adjust it
```

3. click `regenerate steps`

> 4. select the first step that broke

Reviewer: I still don't fully know how to use this feature properly but wouldn't you select the last successful step (and not want to keep the broken one?)

Author: ah, richtig

Author: ```suggestion
4. select the last step you wish to keep
```

5. confirm `generate steps`

> The **AI agent** will is re-generate steps of your test and validate them. If it's alright, click `save & run`. If it needs adjustment, this is [how you edit test cases](/edit-test-case.mdx).

Reviewer: grammar

Author: ```suggestion
The **AI agent** will re-generate steps of your test and validate them. If it's alright, click `save & run`. If it needs adjustment, this is [how you edit test cases](/edit-test-case.mdx).
```

> The **AI agent** will is re-generate steps of your test and validate them. If it's alright, click `save & run`. If it needs adjustment, this is [how you edit test cases](/edit-test-case.mdx).

Reviewer: alright in what sense?

<Tip>

>   If snapshots in preview don't display properly, or the state of the run is all

Reviewer: This sounds a little defensive (as if we would have bugs here ;) so maybe we rather speak about the fact that runs (and related snapshots and traces) go stale/ outdated and that you can always start another run to see an up-to-date state?

Author: wanna join marketing?
